google/cloud/sql/connector/connector.py

```diff
@@ -25,41 +25,125 @@
+
+    def _get_loop(self) -> asyncio.AbstractEventLoop:
+        if self._loop is None:
+            self._loop = asyncio.new_event_loop()
+            self._thread = Thread(target=self._loop.run_forever, daemon=True)
+            self._thread.start()
+        return self._loop
+
+    def _get_keys(self, loop: asyncio.AbstractEventLoop) -> concurrent.futures.Future:
+        if self._keys is None:
+            self._keys = asyncio.run_coroutine_threadsafe(generate_keys(), loop)
+        return self._keys
+
+    def connect(
+        self, instance_connection_string: str, driver: str, **kwargs: Any
+    ) -> Any:
+        """Prepares and returns a database connection object and starts a
+        background thread to refresh the certificates and metadata.
+
+        :type instance_connection_string: str
+        :param instance_connection_string:
+            A string containing the GCP project name, region name, and instance
+            name separated by colons.
+
+            Example: example-proj:example-region-us6:example-instance
+
+        :type driver: str
+        :param: driver:
+            A string representing the driver to connect with. Supported drivers are
+            pymysql, pg8000, and pytds.
+
+        :param kwargs:
+            Pass in any driver-specific arguments needed to connect to the Cloud
+            SQL instance.
+
+        :rtype: Connection
+        :returns:
+            A DB-API connection to the specified Cloud SQL instance.
+        """
+
+        # Initiate event loop and run in background thread.
+        #
+        # Create an InstanceConnectionManager object from the connection string.
+        # The InstanceConnectionManager should verify arguments.
+        #
+        # Use the InstanceConnectionManager to establish an SSL Connection.
+        #
+        # Return a DBAPI connection
+
+        loop = self._get_loop()
+        if instance_connection_string in self._instances:
+            icm = self._instances[instance_connection_string]
+        else:
+            keys = self._get_keys(loop)
+            enable_iam_auth = kwargs.pop("enable_iam_auth", self._enable_iam_auth)
+            icm = InstanceConnectionManager(
+                instance_connection_string, driver, keys, loop, enable_iam_auth
+            )
+            self._instances[instance_connection_string] = icm
+
+        ip_types = kwargs.pop("ip_types", self._ip_types)
+        if "timeout" in kwargs:
+            return icm.connect(driver, ip_types, **kwargs)
+        elif "connect_timeout" in kwargs:
+            timeout = kwargs["connect_timeout"]
+        else:
+            timeout = self._timeout
+        try:
+            return icm.connect(driver, ip_types, timeout, **kwargs)
+        except Exception as e:
+            # with any other exception, we attempt a force refresh, then throw the error
+            icm.force_refresh()
+            raise (e)
+
+
+_default_connector: Optional[Connector] = None
```

Comment:
nit: Let's try to get global state to the top of the file so it's harder to miss